azurelinuxagent/ga/exthandlers.py

```diff
@@ -949,31 +993,37 @@ def launch_command(self, cmd, timeout=300):
         begin_utc = datetime.datetime.utcnow()
         self.logger.verbose("Launch command: [{0}]", cmd)
         base_dir = self.get_base_dir()
+
+        def sanitize(s):
+            return ustr(s, encoding='utf-8', errors='backslashreplace')
+
         try:
-            devnull = open(os.devnull, 'w')
-            child = subprocess.Popen(base_dir + "/" + cmd,
-                                     shell=True,
-                                     cwd=base_dir,
-                                     stdout=devnull,
-                                     env=os.environ)
+            # This should be .run(), but due to the wide variety
+            # of Python versions we must support we must use .communicate().
```

Comment:
Which particular python versions do we support that require us to use `.communicate()`? Some day we will drop support for distros that use those versions and we'll be able to switch to `.run()`, and it would be nice to know when that day has come.